File: src/Muflone/Persistence/IRepository.cs
using System;
using System.Collections.Generic;
using System.Threading;
using System.Threading.Tasks;

namespace Muflone.Persistence;

public interface IRepository : IDisposable
{
	Task<TAggregate> GetByIdAsync<TAggregate>(Guid id, CancellationToken cancellationToken = default) where TAggregate : class, IAggregate;
	Task<TAggregate> GetByIdAsync<TAggregate>(Guid id, long version, CancellationToken cancellationToken = default) where TAggregate : class, IAggregate;
	Task SaveAsync(IAggregate aggregate, Guid commitId, Action<IDictionary<string, object>> updateHeaders, CancellationToken cancellationToken = default);
	Task SaveAsync(IAggregate aggregate, Guid commitId, CancellationToken cancellationToken = default);
}

File: php/src/Muflone/Persistence/IRepository.php
<?php

declare(strict_types=1);

namespace Muflone\Persistence;

use Muflone\Core\IDomainId;
use Muflone\IAggregate;

interface IRepository
{
    /**
     * Get an aggregate by its ID
     *
     * @template T of IAggregate
     * @param class-string<T> $aggregateClass
     * @param IDomainId $id
     * @return T|null
     */
    public function getById(string $aggregateClass, IDomainId $id): ?IAggregate;

    /**
     * Get an aggregate by its ID and version
     *
     * @template T of IAggregate
     * @param class-string<T> $aggregateClass
     * @param IDomainId $id
     * @param int $version
     * @return T|null
     */
    public function getByIdAndVersion(string $aggregateClass, IDomainId $id, int $version): ?IAggregate;

    /**
     * Save an aggregate
     *
     * @param IAggregate $aggregate
     * @param string $commitId
     * @param callable|null $updateHeaders Optional callback to update headers
     * @return void
     */
    public function save(IAggregate $aggregate, string $commitId, ?callable $updateHeaders = null): void;
}
